### ISSUE TITLE
:db/retractEntity failing to retract all component datoms

### ISSUE THREAD

**Issue by gfpeltier** [tonsky/datascript#308]
:db/retractEntity and :db.fn/retractEntity are failing to remove all component datoms from the db. Reproducible example (Clojure 1.9.0 & Datascript 0.18.1):

```clojure
(def pschema {:person/things {:db/valueType :db.type/ref 
                              :db/cardinality :db.cardinality/many 
                              :db/isComponent true}})

(def conn (d/create-conn pschema))

(d/transact! conn [{:person/name "Guy"
                    :person/things [{:thing/name "A"
                                     :thing/type :a}
                                    {:thing/name "B"
                                     :thing/type :a}
                                    {:thing/name "C"
                                     :thing/type :a}]}])


(d/transact! conn [[:db/retractEntity 1]])

@conn
=> #datascript/DB{:schema #:person{:things #:db{:valueType :db.type/ref,
                                             :cardinality :db.cardinality/many,
                                             :isComponent true}},
                  :datoms [[2 :thing/type :a 536870913]]}
```

It seems tangentially related to #294 so I tried to go reproduce this using 0.18.4 however attempting to retract entities there consistently threw exceptions related to persistent-sorted-set:

```clojure
(d/transact! conn [[:db/retractEntity 1]])
RuntimeException Tovarisch, you are iterating and mutating a transient set at the same time!  me.tonsky.persistent_sorted_set.Seq.checkVersion (Seq.java:31)
```

**Comment by tonsky** [comment]
This looks bad! Thanks, I’ll take a look shortly

**Comment by gfpeltier** [comment]
Thank you for such a quick response! I went back and took another look this morning and I think the 0.18.4 issue was just an environment problem on my end. Your fix in 0.18.4 seems to also have fixed the `:db/retractEntity` issue I was seeing in 0.18.1. Sorry for the false alarm.

**Comment by tonsky** [comment]
Oh. Well, I’m glad it’s resolved!
